fs/vfs/fs_uio.c

```diff
@@ -0,0 +1,194 @@
+/****************************************************************************
+ * fs/vfs/fs_uio.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <sys/uio.h>
+#include <sys/types.h>
+
+#include <nuttx/fs/fs.h>
+#include <nuttx/fs/uio.h>
+
+#include <assert.h>
+#include <errno.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: iovec_total_len
+ *
+ * Description:
+ *   Return the total length of data in bytes.
+ *   Or -EOVERFLOW.
+ *
+ ****************************************************************************/
+
+ssize_t iovec_total_len(FAR const struct iovec *iov, int iovcnt)
+{
+  size_t len = 0;
+  int i;
+
+  for (i = 0; i < iovcnt; i++)
+    {
+      if (SSIZE_MAX - len < iov[i].iov_len)
+        {
+          return -EOVERFLOW;
+        }
+
+      len += iov[i].iov_len;
+    }
+
+  return len;
+}
+
+/****************************************************************************
+ * Name: iovec_compat_readv
+ *
+ * Description:
+ *   Emulate readv using file_operation::compat_read.
+ *
+ ****************************************************************************/
+
+ssize_t iovec_compat_readv(FAR struct file *filep,
```

Comment:
could move to fs_read.c as a static function, and call it directly if driver/fs doesn't implement readv